src/Compilers/Core/Portable/InternalUtilities/SetWithInsertionOrder.cs

```diff
@@ -37,6 +37,27 @@ public bool Add(T value)
             return true;
         }
 
+        public bool Insert(int index, T value)
+        {
+            if (_set == null)
+            {
+                if (index > 0)
+                {
+                    throw new IndexOutOfRangeException();
+                }
+                Add(value);
+            }
+            else
+            {
+                if (!_set.Add(value))
+                {
+                    return false;
+                }
+                _elements.Insert(index, value);
```

Comment:
Consider the case the `Insert` operation throws because it's invalid. That means `SetWithInsertionOrder` is now in an inconsistent state: `_set` has the value but it's not present in `_elements`. Think we need to guard against this and ensure the data type is consistent in the face of bad `index` values.